### ISSUE TITLE
Where is  fetching batch data operation from queue  in   training and evaluation code?

### ISSUE THREAD

**Issue by lifematrix** [kwotsin/transfer_learning_tutorial#5]
Hey,  your code help me a  lot, and thanks again.

I can understand ```load_batch``` function in ```train_flowers.py``` will create  a queue to  produce  batch data for training. 

Howerver, in the  loop: 
```
for step in xrange(num_steps_per_epoch * num_epochs):
```
I can not find which operation do the work of fetching another batch explicitly. I guess it occurs in ```train_op``` automatically. Is it right?

For  ```eval_flowers.py```, I have the  same question.

Can you give any explanation?

Thanks!

Steven

**Comment by kwotsin** [comment]
Firstly, your dataset is shuffled when you read them using the `get_split` function, and then when you load your images into one batch using `tf.train.batch`, they are loaded into a FIFO queue. The same function `tf.train.batch` will dequeue the examples automatically from this FIFO queue in order to train your model, so you do not have to worry about explicitly getting the next batch. So each time you run the for-loop in the code, you are dequeueing a new batch of examples. You can refer here for more documentation:

1. https://www.tensorflow.org/api_docs/python/tf/train/batch

2. https://www.tensorflow.org/programmers_guide/supervisor

**Comment by lifematrix** [comment]
Thanks for your reply. For the code  in ```train_flowers.py```:

```

		#Run the managed session
		with sv.managed_session() as sess:
			print("num_steps_per_epoch: %d, num_epochs: %d" % (num_steps_per_epoch, num_epochs))
			for step in xrange(num_steps_per_epoch * num_epochs):
			# for step in xrange(1):
				#At the start of every epoch, show the vital information:
				if step % num_batches_per_epoch == 0:
					logging.info('Epoch %s/%s', step/num_batches_per_epoch + 1, num_epochs)
					learning_rate_value, accuracy_value = sess.run([lr, accuracy])
					logging.info('Current Learning Rate: %s', learning_rate_value)
					logging.info('Current Streaming Accuracy: %s', accuracy_value)

					# optionally, print your logits and predictions for a sanity check that things are going fine.
					logits_value, probabilities_value, predictions_value, labels_value = sess.run([logits, probabilities, predictions, labels])
					print 'logits: \n', logits_value
					print 'Probabilities: \n', probabilities_value
					print 'predictions: \n', predictions_value
					print 'Labels:\n:', labels_value

				#Log the summaries every 10 step.
				if step % 10 == 0:
					loss, _ = train_step(sess, train_op, sv.global_step)
					summaries = sess.run(my_summary_op)
					sv.summary_computed(sess, summaries)
					
				#If not, simply run the training step
				else:
					loss, _ = train_step(sess, train_op, sv.global_step)
```
You mean, in the code block ```with sv.managed_session() as sess:```,  ```sess``` will sense the  for-loop, and fetch a new batch for  each iteration. It seams so magically ：). Really?

**Comment by kwotsin** [comment]
The session doesn't sense the for loop, but rather, use of a for-loop to continuously run the session will let the next batch get dequeued and be used to train the model. Each instance of running one loop of a for loop will call one batch of data to train. I initially didn't know this happens too, as most of the dequeuing operations happen under the hood.

**Comment by lifematrix** [comment]
Thanks for your explanation that really help me understand tensorflow better, I got it! Tensorflow  is so smart and big framework, I should have more time  to dive into it.

Thanks again for your sharing and help.

Steven
